Avoid duplicating complex expression in comparisons

ranma42 opened the pull request:
When comparing a nullable expression to a non-nullable one, a  `NULL` result  always
represent a difference.

This makes it possible to avoid duplicating the nullable expression by mapping
the `NULL` result to a `FALSE` (when comparing for equality).

Fixes #34165.

ranma42 commented:
This change can already take care of most of the worst offenders found in #34048 🥳

ranma42 commented:
I'll add some tests that check this transformation specifically

ranma42 commented on src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:
```diff
@@ -1815,10 +1815,23 @@ or ExpressionType.LessThan
-        body = _sqlExpressionFactory.OrElse(
-            _sqlExpressionFactory.AndAlso(body, _sqlExpressionFactory.AndAlso(leftIsNotNull, rightIsNotNull)),
-            _sqlExpressionFactory.AndAlso(leftIsNull, rightIsNull));
+        if (leftNullable && rightNullable
```
I added [`4c3a542` (#34172)](https://github.com/dotnet/efcore/pull/34172/commits/4c3a542c85b4f0aa610da9ae7a710cdc8492d4ab) which is aimed at addressing this

roji commented on test/EFCore.SqlServer.FunctionalTests/Query/JsonQuerySqlServerTest.cs:
```diff
@@ -1003,7 +1003,10 @@ public override async Task Json_collection_index_in_predicate_using_constant(boo
             """
 SELECT [j].[Id]
 FROM [JsonEntitiesBasic] AS [j]
-WHERE JSON_VALUE([j].[OwnedCollectionRoot], '$[0].Name') <> N'Foo' OR JSON_VALUE([j].[OwnedCollectionRoot], '$[0].Name') IS NULL
```
This is a place where I'm a bit hesitant about this change... The [SQL Server docs](https://learn.microsoft.com/en-us/sql/relational-databases/json/index-json-data) specifically document using an indexed computed column over JSON_VALUE as a way to speed up queries filtering inside a JSON document; unless I'm mistaken, these queries would likely stop using such an index if we switch to the CASE translation (maybe in this specific test it doesn't matter because of the inequality, but you get whar I'm saying).

In a perfect world, we'd vary our translation based on knowledge that an indexed computed column exists for this expression, but we're pretty far away from doing that at the moment.

Thoughts?

ranma42 replied:
Yes, I believe it is very likely that the `CASE` translation will not take advantage of indexes, but I would expect the same to be true for the original version as well, as it is performing a `<>` comparison (maybe it would use the index to include all of the `NULL` values 🤔 ).

For equality in predicates the translation should already be
`WHERE JSON_VALUE([j].[OwnedCollectionRoot], '$[0].Name') = N'Foo'`
which should effectively take advantage of indexes.

roji replied:
OK.

So I'm trying to understand whether there are cases - and which ones - in which this PR causes a perf regression because the switch to CASE doesn't use an index. If there **are** such cases (and after all, we **do** avoid the CASE translation for columns because of this), we should think carefully - I'm not sure whether the optimization to remove double evaluation for some cases outweighs the (potentially severe) regression triggered by not using an index. A conservative approach would wait until we could know more reliably whether an index would be used on an expression (e.g. because we're aware of expression indexes/indexed computed columns).

I know I'm being very cautious here, I'm thinking about the perf regressions brought about by the switch from IN+constants to OPENJSON in 8.0 - that change improved general perf for many queries, but also caused severe regressions for others.

ranma42 replied:
Yes, there are cases in which the translation could cause a regression; the main one I can think of (which is the one currently avoided by the column handling) is the following (and similar ones):

```csharp
.Where(e => !e.BoolA != e.NullableBoolB)
```

This is 

```sql
SELECT "e"."Id"
FROM "Entities1" AS "e"
WHERE "e"."BoolA" = "e"."NullableBoolB" OR "e"."NullableBoolB" IS NULL
```

Sqlite (and litely other SQL providers) would take advantage of an index on `NullableBoolB` (assuming BoolA and NullableBoolB are actually columns from different tables).

When using the `CASE`, this becomes

```sql
SELECT "e"."Id"
FROM "Entities1" AS "e"
WHERE CASE
    WHEN "e"."BoolA" <> "e"."NullableBoolB" THEN 0
    ELSE 1
END
```

and the index cannot be used anymore.

I pushed https://github.com/ranma42/efcore/commit/ecdd12ebe5abca48d54f80ea09bf5ce07d262b86 to show what happens when the `CASE` transformation is used whenever it is valid.

With https://github.com/dotnet/efcore/pull/34166 this could possibly affect a few more tests, but if I am not mistaken, this boolean comparison (negated-different-from) is the only case in which a "good" `WHERE` would regress  (at least according to optimizations rules similar to those of sqlite).

ranma42 replied:
ah, obviously you could also do the same on a json value:
```sql
.Where(e => !e.MyJsonColumn.BoolA != e.NullableBoolB)
```

ranma42 replied:
Maybe instead of checking for a simple column, the right check would be whether the emitted operand is `=`  vs `!=`? (aka if the `WHERE` predicate has some chances of being optimized)

ranma42 committed:
Avoid duplicating complex expression in comparisons

When comparing a nullable expression to a non-nullable one, a  `NULL` result  always
represent a difference.

This makes it possible to avoid duplicating the nullable expression by mapping
the `NULL` result to a `FALSE` (when comparing for equality).

Fixes #34165.

ranma42 committed:
Add test to check for (non-)duplication of equality subtrees

ranma42 committed:
Update baselines

ranma42 commented:
I pushed a new version of the branch to solve the merge conflicts.
As I was at it, I also changed the logic behind the activation of the CASE transformation; it now only activates if it is valid (not on nullable vs nullable) and it causes no de-optimization (aka it is only allowed on predicates if the comparison is an inequality).

ranma42 commented on test/EFCore.SqlServer.FunctionalTests/Query/NorthwindMiscellaneousQuerySqlServerTest.cs:
```diff
@@ -2844,7 +2862,7 @@ public override async Task OrderBy_comparison_operator(bool async)
 SELECT [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
 FROM [Customers] AS [c]
 ORDER BY CASE
-    WHEN [c].[Region] = N'ASK' AND [c].[Region] IS NOT NULL THEN CAST(1 AS bit)
+    WHEN [c].[Region] = N'ASK' THEN CAST(1 AS bit)
```
this is a nice side-effect, but we might want to ensure that this kind of optimization happens regardless of this PR (and possibly not only on comparisons 🤔 )

roji commented on test/EFCore.SqlServer.FunctionalTests/Query/AdHocMiscellaneousQuerySqlServerTest.cs:
```diff
@@ -2251,7 +2251,10 @@ SELECT MIN([o0].[HourlyRate])
     FROM [Order] AS [o0]
     WHERE [o0].[CustomerId] = [o].[CustomerId]) AS [CustomerMinHourlyRate], MIN([o].[HourlyRate]) AS [HourlyRate], COUNT(*) AS [Count]
 FROM [Order] AS [o]
-WHERE [o].[Number] <> N'A1' OR [o].[Number] IS NULL
```
Doesn't this regress performance, at least for the case where Number is NULL? Is it worth making an exception for non-complex expressions, and not do the CASE translation?

roji replied:
Right, similar to this conversation above: https://github.com/dotnet/efcore/pull/34172#pullrequestreview-2203171043

roji commented:
/cc @maumar

roji commented:
/azp run

ranma42 commented:
rebased to resolve conflicts
